vehicle/src/Vehicle/Backend/Isabelle/Compile.hs

```diff
@@ -0,0 +1,897 @@
+import Vehicle.Data.Universe (UniverseLevel (..))
+import Vehicle.Data.Variable.Bound.Context.Name (MonadNameContext, addNameToContext, ixToProperName, runFreshNameBoundContextT, getNameContext)
+import Vehicle.Syntax.Builtin
+import Vehicle.Syntax.Sugar
+  ( BinderType (..),
+    LetBinder,
+    foldBinders,
+    foldDeclBinders,
+  )
+import Vehicle.Syntax.Tensor
+  ( Tensor (..),
+    TensorShape,
+    foldMapTensor,
+    shapeOf,
+    toList,
+  )
+
+--------------------------------------------------------------------------------
+-- Isabelle-specific options
+
+data IsabelleOptions = IsabelleOptions
+  { output :: Maybe FilePath,
+    localeName :: Maybe String
+  }
+
+currentPhase :: Doc ()
+currentPhase = "compilation to Isabelle"
+compileProgToIsabelle :: (MonadCompile m) => Prog DecidabilityBuiltin -> IsabelleOptions -> m (Doc a)
+compileProgToIsabelle (Main ds) options =
+  logCompilerSection2 MinDetail currentPhase $ do
+    logDebug MaxDetail $ prettyExternal (Main ds)
+    -- Combine the printed documents
+    -- let combined = fold gatheredStatements
+
+    -- Extract all locale assumptions (not as Doc annotations)
+    -- let localeAssms = foldMap (getLocaleAssms . annotationOf) gatheredStatements
+    (localeNets, localeAssms, programDoc) <- runFreshNameBoundContextT $ do
+      localeNets <- fmap concat (traverse (gatherLocaleNetworks options) ds)
+      localeAssms <- fmap concat (traverse (gatherLocaleStatements options localeNets) ds)
+      programDoc <- compileProg options (localeNets ++ localeAssms) (Main ds)
+      return (localeNets, localeAssms, programDoc)
+    let programDependencies =
+          collectCodeDependencies programDoc
+            `Set.union` collectLocaleDependencies localeNets
+            `Set.union` collectLocaleDependencies localeAssms
+
+    let nameOfLocale = Text.pack $ case localeName options of
+          Just name -> name
+          _ -> maybe "Spec" takeBaseName (output options)
+
+    let isabelleProgram =
+          unAnnotate
+            ( (vsep2 :: [Code] -> Code)
+                [ preamble nameOfLocale programDependencies localeAssms,
+                  indent 2 programDoc,
+                  postamble nameOfLocale (localeNets ++ localeAssms)
+                ]
+            )
+
+    return isabelleProgram
+
+
+-- | Collect dependencies from a 'Code' document by discarding precedence
+--   and folding all dependency annotations.
+collectCodeDependencies :: Code -> Set Dependency
+collectCodeDependencies doc =
+  let stream = layoutPretty defaultLayoutOptions doc
+   in fold (reAnnotateS fst stream)
+
+-- | Collect dependencies arising from a locale definition. This ensures
+--   that imports required by statements printed in the postamble are
+--   included in the preamble import list.
+collectLocaleDependencies :: [LocaleDef] -> Set Dependency
+collectLocaleDependencies = Set.unions . fmap deps
+  where
+    deps :: LocaleDef -> Set Dependency
+    deps (NetworkDefStatement name ty) =
```

Comment:
Usually cleaner to use `\case`, i.e.
```
deps = \case
    NetworkDefStatement name ty ->  collectCodeDependencies name `Set.union` collectCodeDependencies ty
    PropertyDefStatement stmt -> collectCodeDependencies stmt
    TensorTypeDefStmt _ shape body -> collectCodeDependencies shape `Set.union` collectCodeDependencies body
    IndexTypeDefStmt _ maxI body -> collectCodeDependencies maxI `Set.union` collectCodeDependencies body
```
Then you can rename the definition without having to touch so many lines of code.

Reply:
fixed (for all occurrences)